diff --git a/clang/include/clang/Basic/Attr.td b/clang/include/clang/Basic/Attr.td
index efa64dbe4b51..a14ec41ae10b 100644
--- a/clang/include/clang/Basic/Attr.td
+++ b/clang/include/clang/Basic/Attr.td
@@ -5344,3 +5344,15 @@ def ModularFormat : InheritableAttr {
   let Subjects = SubjectList<[Function]>;
   let Documentation = [ModularFormatDocs];
 }
+
+/*------------------*/
+/* SkePU attributes */
+/*------------------*/
+
+/* [[skepu::userconstant]] */
+def SkepuUserConstant : InheritableAttr
+{
+  let Spellings = [CXX11<"skepu", "userconstant">];
+  let Subjects = SubjectList<[Var]>;
+  let Documentation = [Undocumented];
+}
diff --git a/clang/lib/Sema/SemaDeclAttr.cpp b/clang/lib/Sema/SemaDeclAttr.cpp
index 3f0ea5c8f3eb..bb4133f8ef31 100644
--- a/clang/lib/Sema/SemaDeclAttr.cpp
+++ b/clang/lib/Sema/SemaDeclAttr.cpp
@@ -8033,6 +8033,15 @@ ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D, const ParsedAttr &AL,
   case ParsedAttr::AT_GCCStruct:
     handleGCCStructAttr(S, D, AL);
     break;
+
+	/*-------------------------*/
+	/* SkePU Attributes. */
+	/*-------------------------*/
+
+	/* [[skepu::userconstant]] */
+	case ParsedAttr::AT_SkepuUserConstant:
+		handleSimpleAttribute<SkepuUserConstantAttr>(S, D, AL);
+		break;
   }
 }
 
diff --git a/clang/tools/CMakeLists.txt b/clang/tools/CMakeLists.txt
index afdd613b4ee9..111f916cd255 100644
--- a/clang/tools/CMakeLists.txt
+++ b/clang/tools/CMakeLists.txt
@@ -53,3 +53,6 @@ add_llvm_external_project(clang-tools-extra extra)
 add_clang_subdirectory(libclang)
 
 add_clang_subdirectory(offload-arch)
+
+# Add the skepu-tool subdirectory
+add_clang_subdirectory(skepu-tool)
\ No newline at end of file
diff --git a/clang/tools/skepu-tool b/clang/tools/skepu-tool
new file mode 120000
index 000000000000..45351279b112
--- /dev/null
+++ b/clang/tools/skepu-tool
@@ -0,0 +1 @@
+../../../src/skepu-tool/
\ No newline at end of file
